torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -161,6 +161,33 @@ def reshard(self) -> None:
         if fsdp_param_group := state._fsdp_param_group:
             fsdp_param_group.reshard()
 
+    def unshard(self, async_op: bool = False) -> Optional["UnshardHandle"]:
+        """
+        Unshards the module's parameters by allocating memory and all-gathering
+        the parameters. This method is *not* recursive.
+
+        Args:
+            async_op (bool): If ``True``, then returns a :class:`UnshardHandle`
+                that has a :meth:`wait` method to wait on the unshard op. If
+                ``False``, then returns ``None`` and waits on the handle inside
+                this function.
+
+        .. note:: If ``async_op=True``, then the user does not have to call
+            :meth:`wait` on the returned handle if the unshard op can be waited
```

Comment:
> if the unshard op can be waited on in the module's pre-forward.

wouldn't this always the case? the module's pre-forward would always wait on the pending unshard op on this module itself?

Reply:
Sorry, I meant more like "if waiting on the unshard op in the module's pre-forward _is tolerable_".

There are two reasons why a user may not want to wait until the pre-forward to naturally call `.wait()` / `wait_for_unshard()`:
1. The user wants to use the result of the all-gather earlier (unlikely)
2. If we use PG stashing (i.e. `TORCH_NCCL_AVOID_RECORD_STREAMS=1`), then calling `.wait()` allows the all-gather input to be freed, which saves memory -- the user may want to call `.wait()` earlier to free the memory earlier

Reply:
added some clarification in the doc